### ISSUE TITLE
POST p:http-request gives "MIME type may not contain reserved characters"

### ISSUE THREAD

**Issue by gimsieke** [ndw/xmlcalabash1#290]
This happens first in 1.1.25 (all Saxon versions) and is probably related to the upgrade to httpclient-4.5.5.jar

Running this pipeline:

```xml
<p:declare-step xmlns:p="http://www.w3.org/ns/xproc" 
  xmlns:cx="http://xmlcalabash.com/ns/extensions"
  xmlns:c="http://www.w3.org/ns/xproc-step" version="1.0" name="mystep">

  <p:input port="source" primary="true">
    <p:inline><doc/></p:inline>
  </p:input>
  <p:output port="result" primary="true">
    <p:pipe port="result" step="post"/>
  </p:output>

  <p:import href="http://xmlcalabash.com/extension/steps/library-1.0.xpl"/>
  
  <p:in-scope-names name="vars"/>
  
  <p:template name="http-request">
    <p:input port="template">
      <p:inline>
        <c:request href="http://httpbin.org/post" method="POST" detailed="true">
          <c:multipart content-type="multipart/form-data" boundary="=-=-=-=-=">
            <c:body content-type="application/xml" disposition='form-data; name="data"; filename="test.xml"'>
              {/*}
            </c:body>
          </c:multipart>
        </c:request>
      </p:inline>
    </p:input>
    <p:input port="source">
      <p:pipe step="mystep" port="source"/>
    </p:input>
    <p:input port="parameters">
      <p:pipe step="vars" port="result"/>
    </p:input>
  </p:template>

  <p:http-request name="post"/>
  
  <p:store href="result.json" cx:decode="true">
    <p:input port="source" select="/c:response/c:body"/>
  </p:store>
</p:declare-step>
```

succeeds in all versions up to and including 1.1.24. It fails with

```
ERROR: Pipeline failed: MIME type may not contain reserved characters
```
in 1.1.25 and newer.

**Comment by ndw** [comment]
I believe that it's related to the 4.4.6 version of the httpclient library:

https://github.com/elastic/elasticsearch/issues/26438#issuecomment-325992186

(My `build.gradle` file is only requesting 4.4.5, but I see that 4.4.9 is being pulled in.)

The content type includes the boundary parameter and therefore contains a ";" which is what the 4.4.6+ version is complaining about.

AFAICT, the `MultipartEntityBuilder` is the new/right way to build multipart bodies. However, it appears to have a completely fatal flaw in that there's no provision for providing additional headers for the various parts.

https://stackoverflow.com/questions/46076359/adding-content-id-header-in-multipart-entity

I have no idea what to do with this.

In the short term, if you can force your app to use the 4.4.5 httpclient library, I think the problem will go away.

**Comment by ndw** [comment]
See also: https://stackoverflow.com/questions/55583751/how-do-i-add-additional-headers-to-individual-parts-when-using-multipartentitybu

**Comment by gimsieke** [comment]
I replaced httpclient-4.5.5.jar and httpcore-4.4.9.jar with httpclient-4.5.2.jar and httpcore-4.4.5.jar in the lib directory of the extracted 1.1.22-99 release zip, but then I get:
`Exception in thread "main" java.lang.NoClassDefFoundError: org/apache/http/client/HttpRequestRetryHandler`
So maybe you’re already using some newer httpclient functionality?

**Comment by ndw** [comment]
I made a small amount of progress with the `MultipartEntityBuilder` but it seems fundamentally designed with form submissions in mind. I've just joined the `httpclient-users` mailing list, hopefully I can get some more answers from there.

☹️

**Comment by ndw** [comment]
I believe this is fixed in 1.1.27.

The content-disposition of parts is labeled "form-data" even when it isn't, but I'm hoping this doesn't cause any problems.
